### DISCUSSION TITLE
global_init_lock caused dead lock in single core processor

### DISCUSSION THREAD

**Discussion by easeude** [curl/curl#10012] [General]
Several days ago I upgraded curl lib in our program from 7.78 to 7.84. After that dead lock reported occasionally and the call chain referenced to curl_easy_init. We reviewed the source code changes and found global_init_lock is added for thread safe in the new version. global_init_lock  uses atomic operation like spin-lock which may causes dead lock when multithread preemption happend in single core processor.
Is the above description right ?  Whether should we delete macro "GLOBAL_INIT_IS_THREADSAFE" in  single core processor?

**Comment by bagder** [comment]
1. isn't this a bug and hence it should rather be an issue and not a discussion?
2. how is this related to a single core CPU at all? You can still use threads on a single core
3. we did fix more init lock issues since then, I would urge you to check if 7.86.0 is better/fixed

**Reply by easeude** [reply]
Hi Bagder, I don't know if it is a bug so I want to discuss here.  
The issue is related to single core CPU,  because in multi core CPU the low priority thread ( which calling curl_easy_init) may be preempted by another high priority thread (also calling curl_easy_init),   but the lower one still  has chance to execute to release the lock, so no dead lock happened.  While in single core CPU the higher thread is in dead loop to get the lock and the lower thread can't be scheduled. Seeing the code below.
I checked 7.86 and there is no change.

```
struct Curl_easy *curl_easy_init(void)
{
  CURLcode result;
  struct Curl_easy *data;

  /* Make sure we inited the global SSL stuff */
  global_init_lock();
###   ///// low priority thread gets the lock and is preempted here
  if(!initialized) {
....
  }
  global_init_unlock();
....
}

static inline void curl_simple_lock_lock(curl_simple_lock *lock)
{
  for(;;) {
    if(!atomic_exchange_explicit(lock, true, memory_order_acquire))
      break;
    /* Reduce cache coherency traffic */
    while(atomic_load_explicit(lock, memory_order_relaxed)) {
      /* Reduce load (not mandatory) */
#if defined(__i386__) || defined(__x86_64__)
      __builtin_ia32_pause();
#elif defined(__aarch64__)
      asm volatile("yield" ::: "memory");
#elif defined(HAVE_SCHED_YIELD)
      sched_yield();
#endif
    }
  }
}
```

**Reply by bagder** [reply]
This is why `sched_yield()` or a similar call is done in the `curl_simple_lock_lock()` function. To allow the code that holds the lock to still execute and unlock.

Are you asking because you can produce a dead lock?

**Reply by easeude** [reply]
> This is why `sched_yield()` or a similar call is done in the `curl_simple_lock_lock()` function. To allow the code that holds the lock to still execute and unlock.
> 
> Are you asking because you can produce a dead lock?

Yes,  '__sched_yield' can trigger thread reschedule.  But for x86 '**builtin_ia32_pause**' just generates the pause machine instruction to reduce cpu load but can't trigger reschedule. Is this right?

**Reply by jzakrzewski** [reply]
From what I can see, this is a pretty standard way of implementing a spin lock, so I'd be surprised if that would prevent the scheduler to switch threads. Moreover, even high priority thread cannot just sit there indefinitely and will be preempted at some point to allow other threads to run.
